### DISCUSSION TITLE
Curl 7.80.0 + OpenSSL 3.0.1 + FIPS not sending Client Hello

### DISCUSSION THREAD

**Discussion by biao-billy** [curl/curl#8585] [Q&A] [answered]
Hi all,

After rebuilding my app against Curl 7.80.0 (tried 7.77.0 as well) and OpenSSL 3.0.1, I am not seeing client hello being sent as soon as I enabled FIPS mode via `EVP_default_properties_enable_fips(NULL, 1)`. Everything seems to work fine without enabling FIPS.

Tcpdump shows while `curl_multi_perform` and `curl_multi_wait` loop was running, only TCP 3-way handshake was completed between my app and the server, no TLS handshake messages were followed.

I did not change how libcurl was configured, only added `enable_fips` to OpenSSL 3.0 config.

Any idea what might be happening? And where to look first?

TIA, Billy

**Comment by biao-billy** [comment]
Update: Noticed the "Insufficient randomness" error after enabling tracing:

```
2022-03-11T13:41:28.960-08:00 DEBUG:: == STATE: CONNECTING => PROTOCONNECT handle 0x5622378cf418; line 1982 (connection #0)
2022-03-11T13:41:28.961-08:00 DEBUG:: == Insufficient randomness
2022-03-11T13:41:28.961-08:00 DEBUG:: == multi_done
2022-03-11T13:41:28.961-08:00 DEBUG:: == The cache now contains 0 members
2022-03-11T13:41:28.961-08:00 DEBUG:: == Closing connection 0
2022-03-11T13:41:28.961-08:00 DEBUG:: == Expire cleared (transfer 0x5622378cf418)
```
Which is originated from this function in `lib/vtls/openssl.c` 

```
static CURLcode ossl_seed(struct Curl_easy *data)
{
  /* This might get called before it has been added to a multi handle */
  if(data->multi && data->multi->ssl_seeded)
    return CURLE_OK;

  if(rand_enough()) {
    /* OpenSSL 1.1.0+ should return here */
    if(data->multi)
      data->multi->ssl_seeded = TRUE;
    return CURLE_OK;
  }
#ifdef HAVE_RANDOM_INIT_BY_DEFAULT
  /* with OpenSSL 1.1.0+, a failed RAND_status is a showstopper */
  failf(data, "Insufficient randomness");
  return CURLE_SSL_CONNECT_ERROR;
#else
```
Question is why Curl thinks OpenSSL is 1.1.0+ ?

**Reply by bagder** [reply]
1.1.0+ means 1.1.0 or later. 3.0.1 is later than 1.1.0.

The question you want answered is why your `RAND_status()` returns 0 that makes this error output happen.

**Reply by jay** [reply]
This is unlikely to be a curl issue. OpenSSL 3 alpha did have a problem with RAND_status returning 0 due to lack of implicit initialization but it was fixed in release versions. OP has since filed an issue with OpenSSL at https://github.com/openssl/openssl/issues/17875

**Comment by biao-billy** [comment, marked as answer]
@jay Yes the issue is FIPS module loading, not curl. I am closing the discussions here. Thanks all.
